nvflare/client/flare_agent.py

```diff
@@ -0,0 +1,356 @@
+        heartbeat_timeout=30.0,
+        resend_interval=2.0,
+        max_resends=None,
+        submit_result_timeout=30.0,
+        metric_pipe=None,
+        task_channel_name=PipeChannelName.TASK,
+        metric_channel_name=PipeChannelName.METRIC,
+    ):
+        """Constructor of Flare Agent. The agent is responsible for communicating with the Flare Client Job cell (CJ)
+        to get task and to submit task result.
+
+        Args:
+            pipe: pipe for communication
+            submit_result_timeout: when submitting task result, how long to wait for response from the CJ
+        """
+        flare_decomposers.register()
+        common_decomposers.register()
+
+        self.logger = logging.getLogger(self.__class__.__name__)
+        self.pipe = pipe
+        self.pipe_handler = PipeHandler(
+            pipe=self.pipe,
+            read_interval=read_interval,
+            heartbeat_interval=heartbeat_interval,
+            heartbeat_timeout=heartbeat_timeout,
+            resend_interval=resend_interval,
+            max_resends=max_resends,
+        )
+        self.submit_result_timeout = submit_result_timeout
+        self.task_channel_name = task_channel_name
+        self.metric_channel_name = metric_channel_name
+
+        self.metric_pipe = metric_pipe
+        self.metric_pipe_handler = None
+        if self.metric_pipe:
+            self.metric_pipe_handler = PipeHandler(
+                pipe=self.metric_pipe,
+                read_interval=read_interval,
+                heartbeat_interval=heartbeat_interval,
+                heartbeat_timeout=heartbeat_timeout,
+                resend_interval=resend_interval,
+                max_resends=max_resends,
+            )
+
+        self.current_task = None
+        self.task_lock = threading.Lock()
+        self.asked_to_stop = False
+
+    def start(self):
+        """Start the agent. This method must be called to enable CJ/Agent communication.
+
+        Returns: None
+
+        """
+        self.pipe.open(self.task_channel_name)
+        self.pipe_handler.set_status_cb(self._status_cb, pipe_handler=self.pipe_handler, channel=self.task_channel_name)
+        self.pipe_handler.start()
+
+        if self.metric_pipe:
+            self.metric_pipe.open(self.metric_channel_name)
+            self.metric_pipe_handler.set_status_cb(
+                self._status_cb, pipe_handler=self.metric_pipe_handler, channel=self.metric_channel_name
+            )
+            self.metric_pipe_handler.start()
+
+    def _status_cb(self, msg: Message, pipe_handler: PipeHandler, channel):
+        self.logger.info(f"{channel} pipe status changed to {msg.topic}: {msg.data}")
+        self.asked_to_stop = True
+        pipe_handler.stop()
+
+    def stop(self):
+        """Stop the agent. After this is called, there will be no more communications between CJ and agent.
+
+        Returns: None
+
+        """
+        self.asked_to_stop = True
+        self.pipe_handler.stop()
```

Comment:
To be consistent, we should also stop self.metric_pipe_handler (if it is not None).

Reply:
thanks, will do